Add preview for OIDC group matching

NobodysNightmare opened the pull request:
# Ticket
https://community.openproject.org/wp/58408

# TODOs

* [x] extract into a modal
* [x] hook up modal forward and backwards
* [x] properly handle regexp parsing errors

NobodysNightmare committed:
Add preview for OIDC group matching

The preview will be rendered in a modal window
on top of the remaining UI to indicate that it contains
elements that are not stored to the database (and indicating that
its usage is very optional).

NobodysNightmare committed:
Send CSRF token on non-GET methods

Previously we'd only have sent it for submitting of forms.
Now anything using request will send a CSRF token, if the request method
requires it.

NobodysNightmare commented on frontend/src/app/core/turbo/turbo-requests.service.ts:
```diff
@@ -31,9 +31,12 @@ export class TurboRequestsService {
       'X-Authentication-Scheme': 'Session',
     };
+    if(init.method && !(init.method === 'GET' || init.method === 'HEAD')) {
+      defaultHeaders['X-CSRF-Token'] = (document.querySelector('meta[name="csrf-token"]') as HTMLMetaElement).content;
```
ESLint complains about me not using a `!` assertion.

Two notes on that:

1. The relevant code was taken as is from a few lines below, so it offended that rule before
2. I can't figure out the correct syntax to make it work... maybe my editor doesn't understand the `!` syntax, but I for sure also don't do that

ba1ash replied:
Does eslint want to have something like the following?
```
      const token = document.querySelector<HTMLMetaElement>('meta[name="csrf-token"]')?.content;
      if (token) {
        defaultHeaders['X-CSRF-Token'] = token;
      }
```

NobodysNightmare commented on frontend/src/stimulus/controllers/dynamic/openid-connect/match-preview-dialog.controller.ts:
```diff
@@ -0,0 +1,101 @@
+
+    this.regexpInputTarget.addEventListener('input', () => { this.updateMatchPreview(); });
+    this.groupNamesInputTarget.addEventListener('input', () => { this.updateMatchPreview(); });
+  }
```
ESLint complains that `connect` should return void, not a Promise as done by using the `async` keyword here.

What's surprising me is that I didn't come up with this idea in the first place, but we have a bunch of other `async connect` methods already. Many of them are async, because they also need to access the plugin context of OpenProject.

Presumably this one is always populated when connecting the stimulus controllers and thus there is no huge practical difference. But technically the code seems to be pretty risky: If `updateMatchPreview` was called before the promise resolves, `this.turboRequests` would be undefined and errors would be raised...

Kharonus replied:
I left a question in the `Development` channel.

Kharonus replied:
No one could tell yet, if stimulus actually supports async connect, so I refactored yours to not use it. :D

ba1ash commented on frontend/src/stimulus/controllers/dynamic/openid-connect/group-sync-form.controller.ts:
```diff
@@ -29,18 +29,34 @@
 
 import { Controller } from '@hotwired/stimulus';
+import { MatchPreviewDialogSubmittedEvent } from './match-preview-dialog.controller';
+import MatchPreviewDialogController from './match-preview-dialog.controller';
```
```suggestion
import MatchPreviewDialogController, { MatchPreviewDialogSubmittedEvent } from './match-preview-dialog.controller';
```

ba1ash commented:
I think it makes sense to check `Synchronize groups` in the following case after clicking apply.

https://github.com/user-attachments/assets/cfab2f2c-f54b-4e32-a085-fd64eec042f6

ba1ash commented on frontend/src/stimulus/controllers/dynamic/openid-connect/match-preview-dialog.controller.ts:
```diff
@@ -0,0 +1,101 @@
+export interface MatchPreviewDialogSubmittedEvent {
+  detail:{
+    regularExpressions:string;
+    previewGroups:string;
```
Do we use `previewGroups` data somewhere? If not then why do we have it as a part of event data?

NobodysNightmare replied:
Indeed we don't use it. I was very much experimenting with Stimulus and thinking about what best practices might be worth establishing.

My thinking was that this dialog doesn't care what you do with its data, but it collects two inputs from the user, so it makes sense to also emit those collected inputs in the submitted event. I.e. the submitted event is generic and doesn't care about the receiver.

I guess the practical impact is low:

* You correctly point out, that it's not used; I also don't believe that we will start using it any time soon
* on the other hand, it also doesn't really hurt to have it in the event data (the event contains a lot of information nobody cares about too)

Kharonus replied:
I'd leave it for now, to have the least possible impact for now on this code. :)

NobodysNightmare commented:
> I think it makes sense to check `Synchronize groups` in the following case after clicking apply.
> Peek.2025-08-21.16-22.mp4

Uh, that's a nice find. I overlooked that, because I was only ever looking at an "enabled" UI... I am wondering whether we should hide the button in case that checkbox is disabled. Though the current structure of the UI is pretty wild. The submit buttons are not part of the `*Form` that defines the stimulus controller, but effectively in the DOM they are in scope of the controller, so we could probably hide the extra button from the controller.

Kharonus committed:
Merge branch 'dev' into group-match-preview

Kharonus committed:
[#66132] update group regex tester

- fix linter errors
- use `div` instead of `li` for matches
- change placeholder text and style
- add testing to a banner
- apply general feedback from PR and UX

Kharonus committed:
[#66132] remove extra buttons from submitorcancel form

Kharonus commented:
> > I think it makes sense to check `Synchronize groups` in the following case after clicking apply.
> > Peek.2025-08-21.16-22.mp4
> 
> Uh, that's a nice find. I overlooked that, because I was only ever looking at an "enabled" UI... I am wondering whether we should hide the button in case that checkbox is disabled. Though the current structure of the UI is pretty wild. The submit buttons are not part of the `*Form` that defines the stimulus controller, but effectively in the DOM they are in scope of the controller, so we could probably hide the extra button from the controller.

This one is solved by the final design, as the "Test pattern" button is no longer available in the form button group, but inside a banner instead. Thus I was also able to retract the changes to the standard `SubmitOrCancelForm`.

Can you have another look, @ba1ash

ba1ash reviewed (approved):
:shipit:
